### PR TITLE
Adding support for cargo clippy output

### PR DESCRIPTION
**Pull request by madelaney**
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

This change will add support for parsing `cargo clippy` output from the rust tool chain, as well as a test for the current `cargo clippy` output.

### CODE REVIEW

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+            Matcher fileInfoMatcher = CARGO_CLIPPY_FILE_PATTERN.matcher(line);
+            if (fileInfoMatcher.matches()) {
+                try {
+                    fileInformation.setFile(fileInfoMatcher.group("file"));
+                    fileInformation.setLine(Integer.parseInt(fileInfoMatcher.group("line")));
+                    fileInformation.setColumn(Integer.parseInt(fileInfoMatcher.group("column")));
+                }
+                catch (NumberFormatException ex) {
+                    throw new IllegalArgumentException("Parser cannot parse file line or column", ex);
```
[the remark this answers is not in the document]
Comment (madelaney) [reply src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:99]: Sure, I'll just ignore/swallow the exception.

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        public String getFile() {
+            return this.fileName;
+        }
+
+        /**
+         * Set the file line where the recommendation originated.
+         *
+         * @param fileline
+         *     The line number.
+         */
```
[the remark this answers is not in the document]
Comment (madelaney) [reply src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:183]: Sorry, Java is not my primary language so I was just trying to follow what I could find.

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    /**
+     * Attempt to pull out the recommendation message from the string.
+     *
+     * @param recommendation
+     * @return Returns the recommendation from the clippy output.
+     */
+    private String sanitizeRecommendation(final String recommendation) {
+        final Pattern expression = Pattern.compile("^.+help:(?<message>.+)");
```
[the remark this answers is not in the document]
Comment (madelaney) [reply src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:135]: Thanks, I'll add that.

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        public void setRecommendation(final String recommendation) {
+            this.recommendation = recommendation;
+        }
+
+        /**
+         * Returns the current clippy recommendation.
+         *
+         * @return The recommendation.
+         */
+        public String getRecommendation() {
```
[the remark this answers is not in the document]
Comment (madelaney) [reply src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:212]: I wasn't sure how that would be formatted in the Jenkins UI, but I can try to see what I can do.

[the remark this answers is not in the document]
Comment (madelaney) [reply src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:250]: I believe I have resolved this.

File: .project [hunk 5]
```diff
@@ -38,4 +38,15 @@
 		<nature>net.sourceforge.pmd.eclipse.plugin.pmdNature</nature>
 		<nature>edu.umd.cs.findbugs.plugin.eclipse.findbugsNature</nature>
 	</natures>
+	<filteredResources>
```
Comment (uhafner) [inline .project:41]: Please revert unrelated changes. 

(I need to remove them from the project in a separate PR)

File: .settings/org.eclipse.core.resources.prefs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 eclipse.preferences.version=1
 encoding//src/main/java=UTF-8
+encoding//src/main/resources=UTF-8
```
Comment (uhafner) [inline .settings/org.eclipse.core.resources.prefs:3]: revert

File: .settings/org.eclipse.jdt.core.prefs
Comment (uhafner) [inline .settings/org.eclipse.jdt.core.prefs:1]: revert

File: src/main/java/edu/hm/hafner/analysis/parser/CargoCheckParser.java [hunk 7]
```diff
@@ -15,10 +8,15 @@
 import edu.hm.hafner.analysis.ReaderFactory;
 import edu.hm.hafner.analysis.Report;
 import edu.hm.hafner.analysis.Severity;
+import java.util.Optional;
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/CargoCheckParser.java:11]: Revert, there is no real change?

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        FileInformation fileInformation = new FileInformation();
+
+        while (lookahead.hasNext(CARGO_CLIPP_CONTEXT_CONTINUES)) {
+            final String line = lookahead.next();
+
+            Matcher fileInfoMatcher = CARGO_CLIPPY_FILE_PATTERN.matcher(line);
+            if (fileInfoMatcher.matches()) {
+                try {
+                    fileInformation.setFile(fileInfoMatcher.group("file"));
+                    fileInformation.setLine(Integer.parseInt(fileInfoMatcher.group("line")));
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParser.java:99]: Use `IntegerParser` then you can remove the try-catch

File: src/main/java/edu/hm/hafner/analysis/parser/cargo/Utils.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Calculate the number of times a character occurs in a string.
+     *
+     * @param str
+     *     The string to count the character in.
+     * @param c
+     *     The character to search for.
+     * @return The occurrence count, or zero.
+     */
+    public static int countChar(final String str, final char c) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/cargo/Utils.java:22]: This method already exists: `StringUtils.countMatches`

File: src/test/java/edu/hm/hafner/analysis/parser/cargo/UtilsTest.java
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/cargo/UtilsTest.java:1]: Not needed, `StringUtils.countMatches`

File: src/test/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParserTest.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                .hasColumnEnd(40)
+                .hasSeverity(Severity.ERROR);
+
+        softly.assertThat(report.get(EXPECTED_ISSUE_COUNT - 1))
+                .hasFileName("src/main.rs")
+                .hasMessage("use of `unwrap_or` followed by a function call")
+                .hasCategory("clippy::or_fun_call")
+                .hasSeverity(Severity.WARNING_NORMAL)
+                .hasLineStart(76)
+                .hasColumnStart(53);
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/cargo/CargoClippyParserTest.java:53]: Can you check the new description for one example as well?

**Review by uhafner — CHANGES_REQUESTED** [review]

**Pull request merged**